h2o-core/src/main/java/water/rapids/ast/prims/models/AstFairnessMetrics.java

```diff
@@ -0,0 +1,470 @@
+                vecs[i] = Vec.makeVec(resultCols[i], fr.domains()[protectedColsIdx[i]], Vec.newKey());
+            }
+            for (int i = 0; i < METRICS_COUNT; i++) {
+                vecs[PCOL_COUNT + i] = Vec.makeVec(resultCols[PCOL_COUNT + i], Vec.newKey());
+            }
+            return new Frame(Key.make("fairness_metrics_" + fr._key + "_for_model_"+ model._key), colNames, vecs);
+        }
+
+        public Map<String, Frame> getROCInfo(Model model, Frame fr) {
+            Map<String, Frame> result = new HashMap<>();
+            for (int id = 0; id < maxIndex; id++) {
+                if (_aucs[id] == null) continue;
+                AUC2 auc = new AUC2(_aucs[id]);
+                // Fill TwoDimTable
+                String[] thresholds = new String[auc._nBins];
+                for (int i = 0; i < auc._nBins; i++)
+                    thresholds[i] = Double.toString(auc._ths[i]);
+                AUC2.ThresholdCriterion crits[] = AUC2.ThresholdCriterion.VALUES;
+                String[] colHeaders = new String[crits.length + 2];
+                String[] colHeadersMax = new String[crits.length];
+                String[] types = new String[crits.length + 2];
+                String[] formats = new String[crits.length + 2];
+                colHeaders[0] = "Threshold";
+                types[0] = "double";
+                formats[0] = "%f";
+                int i;
+                for (i = 0; i < crits.length; i++) {
+                    if (colHeadersMax.length > i) colHeadersMax[i] = "max " + crits[i].toString();
+                    colHeaders[i + 1] = crits[i].toString();
+                    types[i + 1] = crits[i]._isInt ? "long" : "double";
+                    formats[i + 1] = crits[i]._isInt ? "%d" : "%f";
+                }
+                colHeaders[i + 1] = "idx";
+                types[i + 1] = "int";
+                formats[i + 1] = "%d";
+                TwoDimTable thresholdsByMetrics = new TwoDimTable("Metrics for Thresholds", "Binomial metrics as a function of classification thresholds", new String[auc._nBins], colHeaders, types, formats, null);
+                for (i = 0; i < auc._nBins; i++) {
+                    int j = 0;
+                    thresholdsByMetrics.set(i, j, Double.valueOf(thresholds[i]));
+                    for (j = 0; j < crits.length; j++) {
+                        double d = crits[j].exec(auc, i); // Note: casts to Object are NOT redundant
+                        thresholdsByMetrics.set(i, 1 + j, crits[j]._isInt ? (Object) ((long) d) : d);
+                    }
+                    thresholdsByMetrics.set(i, 1 + j, i);
+                }
+                String groupName = keyToString(id, fr);
+                Frame f = thresholdsByMetrics.asFrame(Key.make("thresholds_and_metrics_" + groupName + "_for_model_" + model._key + "_for_frame_" + fr._key));
+                DKV.put(f);
+                result.put("thresholds_and_metrics_" + groupName, f);
```

Comment:
Good question, thank you @sebhrusen! I assumed the `ValMapFrame` would behave the same way as `ValFrame` but on the client the implementation seems to be lacking any memory management.

I'm not sure if it's intentional or, as it seems to me, `ValMapFrame` was just used in some initial [implementation of Target Encoding](https://github.com/h2oai/h2o-3/commit/78fe5b91a7577f0f643e9ced521bdc3e241b790e) without dealing with the memory management.

For this use-case, I think it makes most sense to GC the frames the same way as we do temporary frames - once the frame goes out of scope in the client release the memory on the backend.

For this PR, I'd do the following without changing the general logic of handling `ValMapFrame`:

Python - track the frames and let them be GCed
R - convert the frames to `data.frame`s and delete the original frames eagerly

This is mainly because we don't have nice native data structure in python so it's good to use `H2OFrame` but at the same time these frames are generally small and it's faster to move them to the client and then manipulate them than using the `H2OFrame` so in R I'd use the `data.frame` as soon as possible.